content/blog/how-tikv-accesses-data.md

```diff
@@ -0,0 +1,263 @@
+
+When the key is stored in the Data CF and the Write CF, the corresponding timestamp will be added to the Key — `startTS` for the Data CF, and `commitTS` for the Write CF.
+
+Suppose we need to write `a = 1`, first a `startTS`, 10 for example, is obtained from the TSO; then comes the PreWrite stage of Percolator, when data is written in the Lock CF and the Data CF, as shown below:
+
+```
+Lock CF: W a = lock
+Data CF: W a_10 = value
+```
+
+If the PreWrite is successful, 2PC enters the Commit phase. A `commitTS`, 11 for example, is obtained firstly from TSO, and then the write continues as below:
+
+```
+Lock CF: D a
+Write CF: W a_11 = 10
+```
+
+If the Commit is successful, the Data CF and the Write CF contain the record for the Key-Value, while the actual data is stored in the Data CF, and the corresponding `startTS` is recorded in the Write CF.
+
+To read data, a `startTS`, 12 for example, is obtained from the TSO, and then the read is processed as below:
+
+```
+Lock CF: R a
+Write CF: S a_12-> a_11 = 10
+Data CF: R a_10
+```
+
+Below are how the Percolator model reads data:
+
+1. Check to see if there is a lock in the Lock CF. If so, the read fails. If not, seek the latest commit version in the Write CF (11 in our case).
+
+2. Get the corresponding startTS, (10 in our case), and then combine the key and startTS in the Data CF to read the corresponding data.
+
+The above is just a brief introduction to the read and write processes of Percolator, which in reality will be more complicated than this.
+
+### RocksDB
+
+TiKV uses RocksDB as the underlying storage engine. For TiKV, any data is eventually be converted into one or more key-value pairs to be persisted in RocksDB.
+
+Each TiKV contains two RocksDB instances. One of them is used to store Raft Log, which we call Raft RocksDB, and the other is to store the actual data, which we call KV RocksDB.
+
+A TiKV store contains multiple Regions. In Raft RocksDB, we use the Region ID as the prefix of the key, which is combined with Raft Log ID to uniquely identify a Raft Log. For example, if there are two Regions with IDs of 1 and 2, then the Raft Log is stored in RocksDB as below:
+
+```
+1_1-> Log {a = 1}
+1_2-> Log {a = 2}
+...
+1_N-> Log {a = N}
+2_1-> Log {b = 2}
+2_2-> Log {b = 3}
+…
+2_N-> Log {b = N}
+```
+
+
+Because we slice the key according to range, we use the key to save directly, similar to the following:
+
+```
+a-> N
+b-> N
+```
+
+It stores two keys, a and b, but does not use any prefix to distinguish them.
+
+RocksDB supports column families, so it can directly correspond to the column families in Percolator. In TiKV, we use the Default CF in RocksDB to directly correspond to the Data CF in Percolator, and use Lock and Write CFs of the same name.
+
+### PD
+
+Each TiKV reports the information of all Regions to Placement Driver (PD), so that  PD gathers the Region information of the entire cluster. Based on this, a Region routing table is formed as shown below:
```

Comment:
```suggestion
Each TiKV reports the information of all Regions to Placement Driver (PD), so that PD is aware of the Region information for the entire cluster. Based on this, a Region routing table is formed as shown below:
```